File: Fortran/IBM/execute/F2018/cintrop_ts29113/AssumedType/unit_tests/AssumedTypeObj07f.c
/**********************************************************************
*  ===================================================================
*  XL Fortran Test Case                          IBM INTERNAL USE ONLY
*  ===================================================================
*
*  TEST CASE TITLE            : AssumedTypeObj06f
*
*  PROGRAMMER                 : Ren, Jian Gang
*  DATE                       : Apr 14, 2012
*  ORIGIN                     : Linux/AIX Compiler Development,
*                             : IBM Software Solutions China Lab
*
*  PRIMARY FUNCTIONS TESTED   : C-interop Assumed-type object
*                                                   
*  SECONDARY FUNCTIONS TESTED : None 
*
*  DRIVER STANZA              : xlc
*  REQUIRED COMPILER OPTIONS  : 
*
*
*  DESCRIPTION                : Calling a non-BIND(C) procedure from Fortran
*                               where the procedure is defined in C with
*                               assumed-type object.
*                               - Actual arg is of polymorphic object
*                                 including unlimited polymorphic 
*
*1234567890123456789012345678901234567890123456789012345678901234567890*/

#include <stdio.h>

typedef struct dt
{
  int i;
}dt_t;

typedef struct et
{
  int i;
  int j;
}et_t;

typedef struct ft
{
  int i;
  int k;
}ft_t;

typedef struct new
{
  int v;
}new_t;

void c_func_dyn(void* a, void* flag)
{
  int f = *(int *)flag;

  if (f == 1)
  {
    dt_t *d = (struct dt *)a;
    printf("dt: i = %d\n", d->i);
  }
  else if (f == 2)
  {
    et_t *e = (struct et *)a;
    printf("et: i = %d\n", e->i);
    printf("et: j = %d\n", e->j);
  }
  else if (f == 3)
  {
    ft_t *f = (struct ft *)a;
    printf("ft: i = %d\n", f->i);
    printf("ft: k = %d\n", f->k);
  }
  else if (f == 4)
  {
    printf("int: %d\n", *(int *)a);
  }
  else if (f == 5)
  {
    new_t *n = (struct new *)a;
    printf("new: n = %d\n", n->v);
  }
  else
  {
    printf("Error!\n");
  }
  return;
}

File: Fortran/IBM/execute/F2018/cintrop_ts29113/AssumedType/unit_tests/AssumedTypeObj07f.f90
!**********************************************************************
!*  ===================================================================
!*
!*  DATE                       : Apr 14, 2012
!*                             : IBM Software Solutions China Lab
!*
!*  PRIMARY FUNCTIONS TESTED   : C-interop Assumed-type object
!*
!*  SECONDARY FUNCTIONS TESTED : None
!*
!*  REQUIRED COMPILER OPTIONS  :
!*
!*  DESCRIPTION                : Calling a BIND(C) procedure from Fortran
!*                               where the procedure is defined in C with
!*                               assumed-type object.
!*                               - Actual arg is of polymorphic object including
!*                                 unlimited polymorphic
!*
!234567890123456789012345678901234567890123456789012345678901234567890

module m
  use, intrinsic :: iso_c_binding
  implicit none

  type dt
    integer(c_int) i
  end type

  type, extends(dt) :: et
    integer(c_int) j
  end type

  type, extends(dt) :: ft
    integer(c_int) k
  end type

  integer, parameter :: MAKE_DT = 1
  integer, parameter :: MAKE_ET = 2
  integer, parameter :: MAKE_FT = 3

  contains
    function make_type(i)
      class(dt), allocatable :: make_type
      integer i

      if (i == MAKE_DT) then
        allocate(make_type, source=dt(1))
      else if (i == MAKE_ET) then
        allocate(make_type, source=et(2, 2))
      else if (i == MAKE_FT) then
        allocate(make_type, source=ft(3, 3))
      else
        error stop
      endif
    end function
end module

program AssumedTypeObj07f
  use, intrinsic :: iso_c_binding
  use m
  implicit none

  type new_t
    integer(c_int) :: newV = -1
  end type

  class(*), allocatable :: a
  integer :: i = 77
  type(new_t) :: n

  interface
    subroutine c_func_dyn(a, flag) BIND(c)
      use, intrinsic :: iso_c_binding
      implicit none

      TYPE(*) :: a, flag
    end subroutine c_func_dyn
  end interface

  allocate(a, source=make_type(MAKE_DT))
  call sub(a, 1)
  deallocate(a)

  allocate(a, source=make_type(MAKE_ET))
  call sub(a, 2)
  deallocate(a)

  allocate(a, source=make_type(MAKE_FT))
  call sub(a, 3)
  deallocate(a)

  allocate(a, source=i)
  call sub(a, 4)
  deallocate(a)

  allocate(a, source=n)
  call sub(a, 5)
  deallocate(a)

  contains
    subroutine sub(a, flag)
      TYPE(*) :: a
      TYPE(*) :: flag

      call c_func_dyn(a, flag)
    end
end

